src/dfa/dfa.rs

```diff
@@ -0,0 +1,601 @@
+use crate::nfa::nfa::NFA;
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+use std::process::id;
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+struct State(String);
+
+enum Tag {
+    Start(usize),
+    End(usize),
+}
+
+struct Transition {
+    from_state: State,
+    symbol: char,
+    to_state: State,
+    tag: Option<Tag>,
+}
+
+pub(crate) struct DFA {
+    start: State,
+    accept: Vec<State>,
+    states: HashSet<State>,
+    transitions: HashMap<State, HashMap<char, Transition>>, // from_state -> symbol -> to_state
```

Comment:
As discussed offline, there's a performance concern in this part; shall we add a TODO to keep track of the issue?